File: twin2multicloud_backend/src/schemas/deployer_config.py
from pydantic import BaseModel, Field
from typing import Literal, Optional
from datetime import datetime
import json


class DeployerConfigUpdate(BaseModel):
    """Request model for updating deployer configuration."""
    deployer_digital_twin_name: Optional[str] = None
    config_events_json: Optional[str] = None
    config_iot_devices_json: Optional[str] = None
    config_json_validated: Optional[bool] = None
    config_events_validated: Optional[bool] = None
    config_iot_devices_validated: Optional[bool] = None
    # Section 3: L1 Payloads
    payloads_json: Optional[str] = None
    payloads_validated: Optional[bool] = None
    # Section 3: L2 User Functions
    processor_contents: Optional[dict[str, str]] = None
    processor_validated: Optional[dict[str, bool]] = None
    processor_requirements: Optional[dict[str, str]] = None  # {deviceId: requirements.txt}
    event_feedback_content: Optional[str] = None
    event_feedback_validated: Optional[bool] = None
    event_feedback_requirements: Optional[str] = None
    event_action_contents: Optional[dict[str, str]] = None
    event_action_validated: Optional[dict[str, bool]] = None
    event_action_requirements: Optional[dict[str, str]] = None  # {funcName: requirements.txt}
    state_machine_content: Optional[str] = None
    state_machine_validated: Optional[bool] = None
    # Section 2: L4 Hierarchy
    hierarchy_content: Optional[str] = None
    hierarchy_validated: Optional[bool] = None
    # Section 3: L4 Scene
    scene_glb_uploaded: Optional[bool] = None
    scene_config_content: Optional[str] = None
    scene_config_validated: Optional[bool] = None
    # Section 3: L4/L5 User Config
    user_config_content: Optional[str] = None
    user_config_validated: Optional[bool] = None


class DeployerConfigResponse(BaseModel):
    """Response model for deployer configuration."""
    twin_id: str
    twin_state: Optional[str] = None
    deployer_digital_twin_name: Optional[str] = None
    config_events_json: Optional[str] = None
    config_iot_devices_json: Optional[str] = None
    config_json_validated: bool = False
    config_events_validated: bool = False
    config_iot_devices_validated: bool = False
    # Section 3: L1 Payloads
    payloads_json: Optional[str] = None
    payloads_validated: bool = False
    # Section 3: L2 User Functions
    processor_contents: Optional[dict[str, str]] = None
    processor_validated: Optional[dict[str, bool]] = None
    processor_requirements: Optional[dict[str, str]] = None
    event_feedback_content: Optional[str] = None
    event_feedback_validated: bool = False
    event_feedback_requirements: Optional[str] = None
    event_action_contents: Optional[dict[str, str]] = None
    event_action_validated: Optional[dict[str, bool]] = None
    event_action_requirements: Optional[dict[str, str]] = None
    state_machine_content: Optional[str] = None
    state_machine_validated: bool = False
    # Section 2: L4 Hierarchy
    hierarchy_content: Optional[str] = None
    hierarchy_validated: bool = False
    # Section 3: L4 Scene
    scene_glb_uploaded: bool = False
    scene_config_content: Optional[str] = None
    scene_config_validated: bool = False
    # Section 3: L4/L5 User Config
    user_config_content: Optional[str] = None
    user_config_validated: bool = False
    has_config_artifacts: bool = False
    has_l1_payloads: bool = False
    has_l2_artifacts: bool = False
    has_l4_l5_artifacts: bool = False
    validation_summary: dict[str, bool] = Field(default_factory=dict)
    updated_at: Optional[datetime] = None
    
    @classmethod
    def from_db(cls, config, twin_state: Optional[str] = None):
        """Convert DB model to response."""
        if config is None:
            return None
        
        def parse_json_dict(json_str):
            if json_str is None:
                return None
            try:
                return json.loads(json_str)
            except (json.JSONDecodeError, TypeError):
                return None
        
        return cls(
            twin_id=config.twin_id,
            twin_state=twin_state,
            deployer_digital_twin_name=config.deployer_digital_twin_name,
            config_events_json=config.config_events_json,
            config_iot_devices_json=config.config_iot_devices_json,
            config_json_validated=bool(config.config_json_validated),
            config_events_validated=bool(config.config_events_validated),
            config_iot_devices_validated=bool(config.config_iot_devices_validated),
            payloads_json=config.payloads_json,
            payloads_validated=bool(config.payloads_validated),
            # L2 fields (parse JSON strings to dicts)
            processor_contents=parse_json_dict(config.processor_contents),
            processor_validated=parse_json_dict(config.processor_validated),
            processor_requirements=parse_json_dict(config.processor_requirements),
            event_feedback_content=config.event_feedback_content,
            event_feedback_validated=config.event_feedback_validated or False,
            event_feedback_requirements=config.event_feedback_requirements,
            event_action_contents=parse_json_dict(config.event_action_contents),
            event_action_validated=parse_json_dict(config.event_action_validated),
            event_action_requirements=parse_json_dict(config.event_action_requirements),
            state_machine_content=config.state_machine_content,
            state_machine_validated=config.state_machine_validated or False,
            # L4 Hierarchy
            hierarchy_content=config.hierarchy_content,
            hierarchy_validated=config.hierarchy_validated or False,
            # L4 Scene
            scene_glb_uploaded=config.scene_glb_uploaded or False,
            scene_config_content=config.scene_config_content,
            scene_config_validated=config.scene_config_validated or False,
            # L4/L5 User Config
            user_config_content=config.user_config_content,
            user_config_validated=config.user_config_validated or False,
            has_config_artifacts=bool(
                config.deployer_digital_twin_name
                or config.config_events_json
                or config.config_iot_devices_json
            ),
            has_l1_payloads=bool(config.payloads_json),
            has_l2_artifacts=bool(
                config.processor_contents
                or config.event_feedback_content
                or config.event_action_contents
                or config.state_machine_content
            ),
            has_l4_l5_artifacts=bool(
                config.hierarchy_content
                or config.scene_glb_uploaded
                or config.scene_config_content
                or config.user_config_content
            ),
            validation_summary={
                "config": bool(config.config_json_validated),
                "events": bool(config.config_events_validated),
                "iot_devices": bool(config.config_iot_devices_validated),
                "payloads": bool(config.payloads_validated),
                "event_feedback": bool(config.event_feedback_validated),
                "state_machine": bool(config.state_machine_validated),
                "hierarchy": bool(config.hierarchy_validated),
                "scene_config": bool(config.scene_config_validated),
                "user_config": bool(config.user_config_validated),
            },
            updated_at=config.updated_at,
        )


class ConfigValidationRequest(BaseModel):
    """Request model for validating config content."""
    content: str
    provider: Optional[str] = None  # Required for L2 types (aws, azure, google)


class ConfigValidationResponse(BaseModel):
    """Response model for config validation."""
    valid: bool
    message: str
    errors: Optional[list[str]] = None


DeployerConfigSectionId = Literal[
    "configuration",
    "payloads",
    "user_logic",
    "digital_twin_assets",
]


class DeployerConfigArtifact(BaseModel):
    artifact_id: str
    section_id: DeployerConfigSectionId
    label: str
    content: str | None = None
    has_content: bool
    validated: bool
    required: bool = False
    validation_key: str | None = None
    requirements: str | None = None


class DeployerConfigSection(BaseModel):
    section_id: DeployerConfigSectionId
    label: str
    artifacts: list[DeployerConfigArtifact] = Field(default_factory=list)
    has_content: bool
    validated: bool
    missing_required_artifacts: list[str] = Field(default_factory=list)
    invalid_artifacts: list[str] = Field(default_factory=list)


class DeployerConfigReadModelResponse(BaseModel):
    schema_version: str = "deployer-config-read-model.v1"
    twin_id: str
    twin_state: str | None = None
    sections: list[DeployerConfigSection]
    validation_summary: dict[str, bool] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    updated_at: datetime | None = None

    @classmethod
    def from_db(cls, config, twin_state: Optional[str] = None):
        warnings: list[str] = []
        processor_contents = _parse_json_dict(config.processor_contents, "processor_contents", warnings)
        processor_validated = _parse_json_dict(config.processor_validated, "processor_validated", warnings)
        processor_requirements = _parse_json_dict(config.processor_requirements, "processor_requirements", warnings)
        event_action_contents = _parse_json_dict(
            config.event_action_contents,
            "event_action_contents",
            warnings,
        )
        event_action_validated = _parse_json_dict(
            config.event_action_validated,
            "event_action_validated",
            warnings,
        )
        event_action_requirements = _parse_json_dict(
            config.event_action_requirements,
            "event_action_requirements",
            warnings,
        )

        sections = [
            _section(
                "configuration",
                "Deployment Configuration",
                [
                    _artifact(
                        "deployer_digital_twin_name",
                        "configuration",
                        "Digital Twin Name",
                        config.deployer_digital_twin_name,
                        bool(config.config_json_validated),
                        required=True,
                        validation_key="config",
                    ),
                    _artifact(
                        "config_events_json",
                        "configuration",
                        "Event Rules",
                        config.config_events_json,
                        bool(config.config_events_validated),
                        required=True,
                        validation_key="events",
                    ),
                    _artifact(
                        "config_iot_devices_json",
                        "configuration",
                        "IoT Devices",
                        config.config_iot_devices_json,
                        bool(config.config_iot_devices_validated),
                        required=True,
                        validation_key="iot_devices",
                    ),
                ],
            ),
            _section(
                "payloads",
                "L1 Payloads",
                [
                    _artifact(
                        "payloads_json",
                        "payloads",
                        "Payload Definitions",
                        config.payloads_json,
                        bool(config.payloads_validated),
                        validation_key="payloads",
                    )
                ],
            ),
            _section(
                "user_logic",
                "L2 User Logic",
                _keyed_artifacts(
                    processor_contents,
                    processor_validated,
                    processor_requirements,
                    section_id="user_logic",
                    artifact_prefix="processor",
                    label_prefix="Processor",
                    validation_key_prefix="processor",
                )
                + [
                    _artifact(
                        "event_feedback",
                        "user_logic",
                        "Event Feedback",
                        config.event_feedback_content,
                        bool(config.event_feedback_validated),
                        validation_key="event_feedback",
                        requirements=config.event_feedback_requirements,
                    ),
                    _artifact(
                        "state_machine",
                        "user_logic",
                        "State Machine",
                        config.state_machine_content,
                        bool(config.state_machine_validated),
                        validation_key="state_machine",
                    ),
                ]
                + _keyed_artifacts(
                    event_action_contents,
                    event_action_validated,
                    event_action_requirements,
                    section_id="user_logic",
                    artifact_prefix="event_action",
                    label_prefix="Event Action",
                    validation_key_prefix="event_action",
                ),
            ),
            _section(
                "digital_twin_assets",
                "L4/L5 Digital Twin Assets",
                [
                    _artifact(
                        "hierarchy_content",
                        "digital_twin_assets",
                        "Hierarchy",
                        config.hierarchy_content,
                        bool(config.hierarchy_validated),
                        validation_key="hierarchy",
                    ),
                    _artifact(
                        "scene_glb",
                        "digital_twin_assets",
                        "3D Scene GLB",
                        "uploaded" if config.scene_glb_uploaded else None,
                        bool(config.scene_glb_uploaded),
                        validation_key="scene_glb",
                    ),
                    _artifact(
                        "scene_config_content",
                        "digital_twin_assets",
                        "Scene Configuration",
                        config.scene_config_content,
                        bool(config.scene_config_validated),
                        validation_key="scene_config",
                    ),
                    _artifact(
                        "user_config_content",
                        "digital_twin_assets",
                        "User Configuration",
                        config.user_config_content,
                        bool(config.user_config_validated),
                        validation_key="user_config",
                    ),
                ],
            ),
        ]

        return cls(
            twin_id=config.twin_id,
            twin_state=twin_state,
            sections=sections,
            validation_summary={
                artifact.validation_key: artifact.validated
                for section in sections
                for artifact in section.artifacts
                if artifact.validation_key
            },
            warnings=warnings,
            updated_at=config.updated_at,
        )


def _parse_json_dict(value: str | None, field_name: str, warnings: list[str]) -> dict:
    if not value:
        return {}
    try:
        parsed = json.loads(value)
    except (json.JSONDecodeError, TypeError):
        warnings.append(f"{field_name} contains invalid JSON and was treated as empty.")
        return {}
    if not isinstance(parsed, dict):
        warnings.append(f"{field_name} is not an object and was treated as empty.")
        return {}
    return parsed


def _artifact(
    artifact_id: str,
    section_id: DeployerConfigSectionId,
    label: str,
    content: str | None,
    validated: bool,
    *,
    required: bool = False,
    validation_key: str | None = None,
    requirements: str | None = None,
) -> DeployerConfigArtifact:
    has_content = bool(content)
    return DeployerConfigArtifact(
        artifact_id=artifact_id,
        section_id=section_id,
        label=label,
        content=content,
        has_content=has_content,
        validated=validated,
        required=required,
        validation_key=validation_key,
        requirements=requirements,
    )


def _keyed_artifacts(
    contents: dict,
    validated: dict,
    requirements: dict,
    *,
    section_id: DeployerConfigSectionId,
    artifact_prefix: str,
    label_prefix: str,
    validation_key_prefix: str,
) -> list[DeployerConfigArtifact]:
    artifacts = []
    for key in sorted(contents.keys()):
        safe_key = str(key)
        artifacts.append(
            _artifact(
                f"{artifact_prefix}:{safe_key}",
                section_id,
                f"{label_prefix}: {safe_key}",
                str(contents.get(key) or ""),
                bool(validated.get(key)),
                validation_key=f"{validation_key_prefix}:{safe_key}",
                requirements=(
                    str(requirements.get(key))
                    if requirements.get(key) is not None
                    else None
                ),
            )
        )
    return artifacts


def _section(
    section_id: DeployerConfigSectionId,
    label: str,
    artifacts: list[DeployerConfigArtifact],
) -> DeployerConfigSection:
    missing = [
        artifact.artifact_id
        for artifact in artifacts
        if artifact.required and not artifact.has_content
    ]
    invalid = [
        artifact.artifact_id
        for artifact in artifacts
        if artifact.has_content and not artifact.validated
    ]
    return DeployerConfigSection(
        section_id=section_id,
        label=label,
        artifacts=artifacts,
        has_content=any(artifact.has_content for artifact in artifacts),
        validated=bool(artifacts) and not missing and not invalid,
        missing_required_artifacts=missing,
        invalid_artifacts=invalid,
    )

File: twin2multicloud_flutter/lib/models/deployer_config.dart
import 'package:equatable/equatable.dart';

import 'calc_params.dart';
import 'wizard_config_requests.dart';

class DeployerConfigData extends Equatable {
  final String? deployerDigitalTwinName;
  final String? configEventsJson;
  final String? configIotDevicesJson;
  final bool configJsonValidated;
  final bool configEventsValidated;
  final bool configIotDevicesValidated;
  final String? payloadsJson;
  final bool payloadsValidated;
  final Map<String, String> processorContents;
  final Map<String, bool> processorValidated;
  final Map<String, String> processorRequirements;
  final String? eventFeedbackContent;
  final bool eventFeedbackValidated;
  final String? eventFeedbackRequirements;
  final Map<String, String> eventActionContents;
  final Map<String, bool> eventActionValidated;
  final Map<String, String> eventActionRequirements;
  final String? stateMachineContent;
  final bool stateMachineValidated;
  final String? hierarchyContent;
  final bool hierarchyValidated;
  final bool sceneGlbUploaded;
  final String? sceneConfigContent;
  final bool sceneConfigValidated;
  final String? userConfigContent;
  final bool userConfigValidated;

  const DeployerConfigData({
    this.deployerDigitalTwinName,
    this.configEventsJson,
    this.configIotDevicesJson,
    this.configJsonValidated = false,
    this.configEventsValidated = false,
    this.configIotDevicesValidated = false,
    this.payloadsJson,
    this.payloadsValidated = false,
    this.processorContents = const {},
    this.processorValidated = const {},
    this.processorRequirements = const {},
    this.eventFeedbackContent,
    this.eventFeedbackValidated = false,
    this.eventFeedbackRequirements,
    this.eventActionContents = const {},
    this.eventActionValidated = const {},
    this.eventActionRequirements = const {},
    this.stateMachineContent,
    this.stateMachineValidated = false,
    this.hierarchyContent,
    this.hierarchyValidated = false,
    this.sceneGlbUploaded = false,
    this.sceneConfigContent,
    this.sceneConfigValidated = false,
    this.userConfigContent,
    this.userConfigValidated = false,
  });

  factory DeployerConfigData.fromJson(Map<String, dynamic> json) {
    return DeployerConfigData(
      deployerDigitalTwinName: _string(
        json['deployer_digital_twin_name'],
        'deployer_digital_twin_name',
      ),
      configEventsJson: _string(
        json['config_events_json'],
        'config_events_json',
      ),
      configIotDevicesJson: _string(
        json['config_iot_devices_json'],
        'config_iot_devices_json',
      ),
      configJsonValidated: _bool(
        json['config_json_validated'],
        'config_json_validated',
      ),
      configEventsValidated: _bool(
        json['config_events_validated'],
        'config_events_validated',
      ),
      configIotDevicesValidated: _bool(
        json['config_iot_devices_validated'],
        'config_iot_devices_validated',
      ),
      payloadsJson: _string(json['payloads_json'], 'payloads_json'),
      payloadsValidated: _bool(
        json['payloads_validated'],
        'payloads_validated',
      ),
      processorContents: _stringMap(
        json['processor_contents'],
        'processor_contents',
      ),
      processorValidated: _boolMap(
        json['processor_validated'],
        'processor_validated',
      ),
      processorRequirements: _stringMap(
        json['processor_requirements'],
        'processor_requirements',
      ),
      eventFeedbackContent: _string(
        json['event_feedback_content'],
        'event_feedback_content',
      ),
      eventFeedbackValidated: _bool(
        json['event_feedback_validated'],
        'event_feedback_validated',
      ),
      eventFeedbackRequirements: _string(
        json['event_feedback_requirements'],
        'event_feedback_requirements',
      ),
      eventActionContents: _stringMap(
        json['event_action_contents'],
        'event_action_contents',
      ),
      eventActionValidated: _boolMap(
        json['event_action_validated'],
        'event_action_validated',
      ),
      eventActionRequirements: _stringMap(
        json['event_action_requirements'],
        'event_action_requirements',
      ),
      stateMachineContent: _string(
        json['state_machine_content'],
        'state_machine_content',
      ),
      stateMachineValidated: _bool(
        json['state_machine_validated'],
        'state_machine_validated',
      ),
      hierarchyContent: _string(json['hierarchy_content'], 'hierarchy_content'),
      hierarchyValidated: _bool(
        json['hierarchy_validated'],
        'hierarchy_validated',
      ),
      sceneGlbUploaded: _bool(json['scene_glb_uploaded'], 'scene_glb_uploaded'),
      sceneConfigContent: _string(
        json['scene_config_content'],
        'scene_config_content',
      ),
      sceneConfigValidated: _bool(
        json['scene_config_validated'],
        'scene_config_validated',
      ),
      userConfigContent: _string(
        json['user_config_content'],
        'user_config_content',
      ),
      userConfigValidated: _bool(
        json['user_config_validated'],
        'user_config_validated',
      ),
    );
  }

  DeployerConfigUpdateRequest toUpdateRequest() => DeployerConfigUpdateRequest(
    deployerDigitalTwinName: deployerDigitalTwinName,
    configEventsJson: configEventsJson,
    configIotDevicesJson: configIotDevicesJson,
    configJsonValidated: configJsonValidated,
    configEventsValidated: configEventsValidated,
    configIotDevicesValidated: configIotDevicesValidated,
    payloadsJson: payloadsJson,
    payloadsValidated: payloadsValidated,
    processorContents: processorContents,
    processorValidated: processorValidated,
    processorRequirements: processorRequirements,
    eventFeedbackContent: eventFeedbackContent,
    eventFeedbackValidated: eventFeedbackValidated,
    eventFeedbackRequirements: eventFeedbackRequirements,
    eventActionContents: eventActionContents,
    eventActionValidated: eventActionValidated,
    eventActionRequirements: eventActionRequirements,
    stateMachineContent: stateMachineContent,
    stateMachineValidated: stateMachineValidated,
    hierarchyContent: hierarchyContent,
    hierarchyValidated: hierarchyValidated,
    sceneGlbUploaded: sceneGlbUploaded,
    sceneConfigContent: sceneConfigContent,
    sceneConfigValidated: sceneConfigValidated,
    userConfigContent: userConfigContent,
    userConfigValidated: userConfigValidated,
  );

  @override
  List<Object?> get props => [
    deployerDigitalTwinName,
    configEventsJson,
    configIotDevicesJson,
    configJsonValidated,
    configEventsValidated,
    configIotDevicesValidated,
    payloadsJson,
    payloadsValidated,
    processorContents,
    processorValidated,
    processorRequirements,
    eventFeedbackContent,
    eventFeedbackValidated,
    eventFeedbackRequirements,
    eventActionContents,
    eventActionValidated,
    eventActionRequirements,
    stateMachineContent,
    stateMachineValidated,
    hierarchyContent,
    hierarchyValidated,
    sceneGlbUploaded,
    sceneConfigContent,
    sceneConfigValidated,
    userConfigContent,
    userConfigValidated,
  ];
}

class DeployerConfigRequirements extends Equatable {
  final List<String> deviceIds;
  final List<String> eventActionNames;
  final bool eventFeedbackRequired;
  final bool eventActionsRequired;
  final bool stateMachineRequired;
  final bool hierarchyRequired;
  final bool sceneRequired;
  final bool userConfigRequired;

  const DeployerConfigRequirements({
    this.deviceIds = const [],
    this.eventActionNames = const [],
    this.eventFeedbackRequired = false,
    this.eventActionsRequired = false,
    this.stateMachineRequired = false,
    this.hierarchyRequired = false,
    this.sceneRequired = false,
    this.userConfigRequired = false,
  });

  factory DeployerConfigRequirements.fromContext({
    required CalcParams? calcParams,
    required String? layer4Provider,
    required String? layer5Provider,
    required List<String> deviceIds,
    required List<String> eventActionNames,
  }) {
    final l4UsesManagedAssets = _isAwsOrAzure(layer4Provider);
    return DeployerConfigRequirements(
      deviceIds: List.unmodifiable(deviceIds),
      eventActionNames: List.unmodifiable(eventActionNames),
      eventFeedbackRequired: calcParams?.returnFeedbackToDevice == true,
      eventActionsRequired: calcParams?.useEventChecking == true,
      stateMachineRequired: calcParams?.triggerNotificationWorkflow == true,
      hierarchyRequired: l4UsesManagedAssets,
      sceneRequired: calcParams?.needs3DModel == true && l4UsesManagedAssets,
      userConfigRequired: _isAwsOrAzure(layer5Provider),
    );
  }

  @override
  List<Object?> get props => [
    deviceIds,
    eventActionNames,
    eventFeedbackRequired,
    eventActionsRequired,
    stateMachineRequired,
    hierarchyRequired,
    sceneRequired,
    userConfigRequired,
  ];
}

enum DeployerArtifactSource { generated, userAuthored }

class DeployerArtifactReadiness extends Equatable {
  final String id;
  final String label;
  final DeployerArtifactSource source;
  final bool required;
  final bool hasContent;
  final bool validated;
  final String? dependency;

  const DeployerArtifactReadiness({
    required this.id,
    required this.label,
    required this.source,
    required this.required,
    required this.hasContent,
    required this.validated,
    this.dependency,
  });

  bool get ready => !required || (hasContent && validated);

  @override
  List<Object?> get props => [
    id,
    label,
    source,
    required,
    hasContent,
    validated,
    dependency,
  ];
}

enum DeployerSectionId { configuration, payloads, userLogic, digitalTwinAssets }

class DeployerSectionReadiness extends Equatable {
  final DeployerSectionId id;
  final String label;
  final List<DeployerArtifactReadiness> artifacts;

  const DeployerSectionReadiness({
    required this.id,
    required this.label,
    required this.artifacts,
  });

  bool get ready => artifacts.every((artifact) => artifact.ready);
  List<String> get missingArtifactIds => artifacts
      .where((artifact) => artifact.required && !artifact.hasContent)
      .map((artifact) => artifact.id)
      .toList(growable: false);
  List<String> get invalidArtifactIds => artifacts
      .where(
        (artifact) =>
            artifact.required && artifact.hasContent && !artifact.validated,
      )
      .map((artifact) => artifact.id)
      .toList(growable: false);

  @override
  List<Object?> get props => [id, label, artifacts];
}

class DeployerConfigReadiness extends Equatable {
  final List<DeployerSectionReadiness> sections;

  const DeployerConfigReadiness({required this.sections});

  factory DeployerConfigReadiness.fromData({
    required DeployerConfigData data,
    required DeployerConfigRequirements requirements,
  }) {
    const generated = DeployerArtifactSource.generated;
    const authored = DeployerArtifactSource.userAuthored;
    bool has(String? value) => value?.trim().isNotEmpty == true;
    DeployerArtifactReadiness artifact(
      String id,
      String label,
      DeployerArtifactSource source,
      bool required,
      bool hasContent,
      bool validated, {
      String? dependency,
    }) => DeployerArtifactReadiness(
      id: id,
      label: label,
      source: source,
      required: required,
      hasContent: hasContent,
      validated: validated,
      dependency: dependency,
    );

    return DeployerConfigReadiness(
      sections: [
        DeployerSectionReadiness(
          id: DeployerSectionId.configuration,
          label: 'Deployment Configuration',
          artifacts: [
            artifact(
              'config:core',
              'Core Configuration',
              generated,
              true,
              has(data.deployerDigitalTwinName),
              data.configJsonValidated,
            ),
            artifact(
              'config:events',
              'Event Rules',
              authored,
              true,
              has(data.configEventsJson),
              data.configEventsValidated,
            ),
            artifact(
              'config:iot-devices',
              'IoT Devices',
              authored,
              true,
              has(data.configIotDevicesJson),
              data.configIotDevicesValidated,
            ),
          ],
        ),
        DeployerSectionReadiness(
          id: DeployerSectionId.payloads,
          label: 'L1 Payloads',
          artifacts: [
            artifact(
              'payloads',
              'Payload Definitions',
              authored,
              true,
              has(data.payloadsJson),
              data.payloadsValidated,
            ),
          ],
        ),
        DeployerSectionReadiness(
          id: DeployerSectionId.userLogic,
          label: 'L2 User Logic',
          artifacts: [
            for (final deviceId in requirements.deviceIds)
              artifact(
                'processor:$deviceId',
                'Processor: $deviceId',
                authored,
                true,
                has(data.processorContents[deviceId]),
                data.processorValidated[deviceId] == true,
                dependency: 'config:iot-devices',
              ),
            artifact(
              'event-feedback',
              'Event Feedback',
              authored,
              requirements.eventFeedbackRequired,
              has(data.eventFeedbackContent),
              data.eventFeedbackValidated,
              dependency: 'config:iot-devices',
            ),
            for (final action in requirements.eventActionNames)
              artifact(
                'event-action:$action',
                'Event Action: $action',
                authored,
                requirements.eventActionsRequired,
                has(data.eventActionContents[action]),
                data.eventActionValidated[action] == true,
                dependency: 'config:events',
              ),
            artifact(
              'state-machine',
              'State Machine',
              authored,
              requirements.stateMachineRequired,
              has(data.stateMachineContent),
              data.stateMachineValidated,
              dependency: 'config:iot-devices',
            ),
          ],
        ),
        DeployerSectionReadiness(
          id: DeployerSectionId.digitalTwinAssets,
          label: 'L4/L5 Digital Twin Assets',
          artifacts: [
            artifact(
              'hierarchy',
              'Hierarchy',
              authored,
              requirements.hierarchyRequired,
              has(data.hierarchyContent),
              data.hierarchyValidated,
            ),
            artifact(
              'scene-config',
              'Scene Configuration',
              authored,
              requirements.sceneRequired,
              has(data.sceneConfigContent),
              data.sceneConfigValidated,
              dependency: 'hierarchy',
            ),
            artifact(
              'scene-glb',
              '3D Scene GLB',
              authored,
              requirements.sceneRequired,
              data.sceneGlbUploaded,
              data.sceneGlbUploaded,
              dependency: 'hierarchy',
            ),
            artifact(
              'user-config',
              'User Configuration',
              authored,
              requirements.userConfigRequired,
              has(data.userConfigContent),
              data.userConfigValidated,
            ),
          ],
        ),
      ],
    );
  }

  DeployerSectionReadiness section(DeployerSectionId id) =>
      sections.firstWhere((section) => section.id == id);

  bool get configurationReady => section(DeployerSectionId.configuration).ready;
  bool get deploymentArtifactsReady => sections
      .where((section) => section.id != DeployerSectionId.configuration)
      .every((section) => section.ready);
  bool get ready => configurationReady && deploymentArtifactsReady;

  @override
  List<Object?> get props => [sections];
}

bool _isAwsOrAzure(String? provider) {
  final normalized = provider?.trim().toUpperCase();
  return normalized == 'AWS' || normalized == 'AZURE';
}

String? _string(Object? value, String field) {
  if (value == null) return null;
  if (value is! String) {
    throw FormatException('Invalid API contract: $field must be a string.');
  }
  return value;
}

bool _bool(Object? value, String field) {
  if (value == null) return false;
  if (value is! bool) {
    throw FormatException('Invalid API contract: $field must be a boolean.');
  }
  return value;
}

Map<String, String> _stringMap(dynamic value, String field) {
  if (value == null) return const {};
  if (value is! Map) {
    throw FormatException('Invalid API contract: $field must be an object.');
  }
  return Map.unmodifiable(
    value.map((key, item) {
      if (key is! String || item is! String) {
        throw FormatException(
          'Invalid API contract: $field must contain string entries.',
        );
      }
      return MapEntry(key, item);
    }),
  );
}

Map<String, bool> _boolMap(dynamic value, String field) {
  if (value == null) return const {};
  if (value is! Map) {
    throw FormatException('Invalid API contract: $field must be an object.');
  }
  return Map.unmodifiable(
    value.map((key, item) {
      if (key is! String || item is! bool) {
        throw FormatException(
          'Invalid API contract: $field must contain boolean entries.',
        );
      }
      return MapEntry(key, item);
    }),
  );
}
